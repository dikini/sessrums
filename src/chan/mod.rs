//! Channel implementations for session types.
//!
//! This module will contain the channel implementations that allow
//! communication according to the protocols defined in the `proto` module.
//! It will be implemented in future tasks.