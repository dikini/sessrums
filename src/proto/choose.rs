//! Choose protocol type for session types.
//!
//! This module defines the `Choose<L, R>` protocol type, which represents a protocol
//! that chooses between two continuations, `L` and `R`.

use std::marker::PhantomData;
use super::Protocol;
use super::offer::Offer;

/// A protocol that chooses between two continuations, `L` and `R`.
///
/// The `Choose<L, R>` type represents a protocol that chooses between
/// two continuations, `L` and `R`. The choice is made by this party.
///
/// # Duality
///
/// The dual of `Choose<L, R>` is `Offer<L::Dual, R::Dual>`. This reflects the fundamental
/// session type principle that if one party chooses, the other party must offer the options.
///
/// This duality relationship ensures protocol compatibility between communicating parties:
/// - When one process chooses between protocols L and R, the other process must
///   offer the dual protocols L::Dual and R::Dual.
/// - The duality relationship is symmetric: `Choose<L, R>::Dual::Dual == Choose<L, R>`.
///
/// ## Duality Transformation
///
/// The duality transformation for Choose follows these rules:
/// 1. Replace Choose with Offer
/// 2. Apply the duality transformation to each branch (L becomes L::Dual, R becomes R::Dual)
///
/// For example:
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Offer, Choose};
///
/// // Original protocol
/// type MyChoose = Choose<Send<i32, End>, Recv<String, End>>;
///
/// // Its dual
/// type MyChooseDual = Offer<Recv<i32, End>, Send<String, End>>;
/// ```
///
/// ## Nested Duality
///
/// The duality relationship extends to nested Choose types:
///
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Offer, Choose};
///
/// // A protocol with nested Choose
/// type NestedChoose = Choose<
///     Choose<Send<i32, End>, Recv<bool, End>>,
///     Send<String, End>
/// >;
///
/// // Its dual has nested Offer
/// type NestedChooseDual = Offer<
///     Offer<Recv<i32, End>, Send<bool, End>>,
///     Recv<String, End>
/// >;
/// ```
///
/// # Type Parameters
///
/// * `L` - The first continuation protocol type. Must implement the `Protocol` trait.
/// * `R` - The second continuation protocol type. Must implement the `Protocol` trait.
///
/// # Examples
///
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Choose};
///
/// // A protocol that chooses between:
/// // 1. Sending an i32 and then ending, or
/// // 2. Receiving a String and then ending
/// type MyProtocol = Choose<Send<i32, End>, Recv<String, End>>;
///
/// // The dual protocol would offer:
/// // 1. Receiving an i32 and then ending, or
/// // 2. Sending a String and then ending
/// // (This would be: Offer<Recv<i32, End>, Send<String, End>>)
/// ```
///
/// # Protocol Composition
///
/// `Choose<L, R>` can be composed with other protocol types to create more complex
/// communication patterns:
///
/// ```
/// use sessrums::proto::{Protocol, Send, Recv, End, Choose};
///
/// // A protocol that first sends a boolean, then chooses between
/// // sending an i32 or receiving a String
/// type ComplexProtocol = Send<bool, Choose<Send<i32, End>, Recv<String, End>>>;
/// ```
pub struct Choose<L, R> {
    _marker: PhantomData<(L, R)>,
}

impl<L: Protocol, R: Protocol> Protocol for Choose<L, R> {
    /// The dual of `Choose<L, R>` is `Offer<L::Dual, R::Dual>`.
    ///
    /// This reflects the fact that if one party chooses, the other party
    /// must offer the options.
    type Dual = Offer<L::Dual, R::Dual>;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::proto::{Send, Recv, End};

    #[test]
    fn test_choose_protocol_implementation() {
        // This test verifies that Choose<L, R> implements Protocol
        
        // We can't directly test the equality of types in Rust,
        // but we can verify that Choose<L, R> implements Protocol
        fn assert_implements_protocol<T: Protocol>() {}
        
        // This will compile only if Choose<Send<i32, End>, Recv<String, End>> implements Protocol
        assert_implements_protocol::<Choose<Send<i32, End>, Recv<String, End>>>();
    }

    #[test]
    fn test_choose_duality() {
        // This test verifies the duality relationship for Choose<L, R>
        
        // Define some simple protocol types for testing
        type ChooseProtocol = Choose<Send<i32, End>, Recv<String, End>>;
        type ExpectedDual = Offer<Recv<i32, End>, Send<String, End>>;
        
        // Check that ChooseProtocol::Dual is ExpectedDual
        fn check_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if ChooseProtocol::Dual is ExpectedDual
        check_dual::<ChooseProtocol, ExpectedDual>();
    }

    #[test]
    fn test_choose_duality_symmetry() {
        // This test verifies the symmetry of the duality relationship for Choose<L, R>
        // Choose<L, R>::Dual is Offer<L::Dual, R::Dual>
        // Offer<L::Dual, R::Dual>::Dual should be Choose<L, R>
        
        // Define some simple protocol types for testing
        type ChooseProtocol = Choose<Send<i32, End>, Recv<String, End>>;
        type DualProtocol = Offer<Recv<i32, End>, Send<String, End>>;
        
        // Check that the duality relationship is symmetric
        fn check_duality_symmetry<P, D>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if the duality relationship is correctly implemented
        check_duality_symmetry::<ChooseProtocol, DualProtocol>();
    }

    #[test]
    fn test_choose_with_complex_composition() {
        // Test Choose in a more complex protocol composition
        
        // A protocol that sends a boolean, then chooses between
        // sending an i32 or receiving a String
        type ComplexProtocol = Send<bool, Choose<Send<i32, End>, Recv<String, End>>>;
        
        // The dual should be Recv<bool, Offer<Recv<i32, End>, Send<String, End>>>
        type ExpectedDual = Recv<bool, Offer<Recv<i32, End>, Send<String, End>>>;
        
        // Check that the dual relationships are correctly established
        fn check_complex_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if ComplexProtocol's dual is correctly derived
        check_complex_dual::<ComplexProtocol, ExpectedDual>();
    }

    #[test]
    fn test_nested_choose_duality() {
        // Test duality with nested Choose types
        
        // A protocol that chooses between:
        // 1. Choosing between Send<i32, End> and Recv<bool, End>
        // 2. Sending a String and then ending
        type NestedChooseProtocol = Choose<
            Choose<Send<i32, End>, Recv<bool, End>>,
            Send<String, End>
        >;
        
        // The dual should be Offer between:
        // 1. Offering between Recv<i32, End> and Send<bool, End>
        // 2. Receiving a String and then ending
        type ExpectedDual = Offer<
            Offer<Recv<i32, End>, Send<bool, End>>,
            Recv<String, End>
        >;
        
        // Check that the dual relationships are correctly established
        fn check_nested_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if NestedChooseProtocol's dual is correctly derived
        check_nested_dual::<NestedChooseProtocol, ExpectedDual>();
    }

    #[test]
    fn test_multiple_level_duality() {
        // Test multiple levels of duality (dual of dual of dual)
        
        // Start with a simple protocol
        type OriginalProtocol = Choose<Send<i32, End>, Recv<String, End>>;
        
        // Get its dual
        type FirstDual = <OriginalProtocol as Protocol>::Dual;
        
        // Get the dual of the dual
        type SecondDual = <FirstDual as Protocol>::Dual;
        
        // Get the dual of the dual of the dual
        type ThirdDual = <SecondDual as Protocol>::Dual;
        
        // The dual of the dual should be the original protocol
        fn check_dual_of_dual<P, D, DD>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = DD>,
            DD: Protocol,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if the dual of the dual is the original protocol
        check_dual_of_dual::<OriginalProtocol, FirstDual, SecondDual>();
        
        // The dual of the dual of the dual should be the dual of the original protocol
        fn check_dual_of_dual_of_dual<P, D, DD, DDD>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = DD>,
            DD: Protocol<Dual = DDD>,
            DDD: Protocol,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if the dual of the dual of the dual is the dual of the original protocol
        check_dual_of_dual_of_dual::<OriginalProtocol, FirstDual, SecondDual, ThirdDual>();
    }

    #[test]
    fn test_mixed_protocol_composition() {
        // Test a mixed composition of Choose and other protocol types
        
        // A protocol with Choose nested within other protocol types
        type MixedProtocol = Send<bool,
            Choose<
                Recv<i32, End>,
                Send<String,
                    Choose<
                        Recv<char, End>,
                        Recv<u8, End>
                    >
                >
            >
        >;
        
        // The dual should have Send/Recv swapped and Choose/Offer swapped
        type ExpectedDual = Recv<bool,
            Offer<
                Send<i32, End>,
                Recv<String,
                    Offer<
                        Send<char, End>,
                        Send<u8, End>
                    >
                >
            >
        >;
        
        // Check that the dual relationships are correctly established
        fn check_mixed_dual<P, D>()
        where
            P: Protocol<Dual = D>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if MixedProtocol's dual is correctly derived
        check_mixed_dual::<MixedProtocol, ExpectedDual>();
    }
}