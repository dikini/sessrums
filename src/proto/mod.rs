//! Protocol type definitions for session types.
//!
//! This module contains the core protocol type definitions used to express
//! communication protocols at the type level.

mod proto;
pub use proto::Protocol;

mod send;
pub use send::Send;

mod recv;
pub use recv::Recv;

mod end;
pub use end::End;

// Phase 2 implementations
mod offer;
pub use offer::Offer;

mod choose;
// Choose will be fully implemented in Task 2.4
// pub use choose::Choose;

// These will be uncommented in Phase 5
// mod rec;
// pub use rec::Rec;

// mod var;
// pub use var::Var;