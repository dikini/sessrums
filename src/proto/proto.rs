//! Core Protocol trait for session types.
//!
//! This module defines the fundamental `Protocol` trait that serves as the
//! foundation for the session type system. Session types allow for compile-time
//! verification of communication protocol adherence between communicating parties.

/// The `Protocol` trait represents a communication protocol type.
///
/// In session type theory, protocols describe the communication behavior between
/// two parties. Each protocol has a dual counterpart that represents the behavior
/// of the other party in the communication.
///
/// # Duality
///
/// Duality is a fundamental concept in session types. For every protocol `P`,
/// there exists a dual protocol `P::Dual` that represents the complementary
/// behavior. For example:
///
/// - If `P` sends a message of type `T`, then `P::Dual` receives a message of type `T`.
/// - If `P` offers a choice between protocols, then `P::Dual` makes a choice between the dual protocols.
///
/// This duality ensures that when two parties follow dual protocols, their
/// communication is guaranteed to be compatible and free from communication errors
/// like deadlocks or protocol violations.
///
/// # Examples
///
/// ```
/// use sez::proto::{Protocol, Send, Recv, End};
///
/// // A protocol that sends an i32, then receives a String, then ends
/// type MyProtocol = Send<i32, Recv<String, End>>;
///
/// // The dual protocol receives an i32, then sends a String, then ends
/// // (This will be implemented in future tasks)
/// // type MyDualProtocol = <MyProtocol as Protocol>::Dual;
/// // This would be equivalent to: Recv<i32, Send<String, End>>
/// ```
///
/// # Type Safety
///
/// The `Protocol` trait and its implementations leverage Rust's type system to
/// ensure protocol adherence at compile time. This means that protocol violations
/// are caught as type errors during compilation, preventing runtime communication
/// errors.
pub trait Protocol: Sized {
    /// The dual protocol type.
    ///
    /// For any protocol `P`, `P::Dual` represents the complementary protocol
    /// that can communicate with `P` without errors.
    ///
    /// The duality relationship is symmetric, meaning that for any protocol `P`:
    /// `P::Dual::Dual == P`
    type Dual: Protocol;
}

#[cfg(test)]
mod tests {
    use super::*;

    // Define some simple protocol types for testing
    struct TestProtocol;
    struct DualTestProtocol;

    // Implement Protocol for TestProtocol
    impl Protocol for TestProtocol {
        type Dual = DualTestProtocol;
    }

    // Implement Protocol for DualTestProtocol
    impl Protocol for DualTestProtocol {
        type Dual = TestProtocol;
    }

    #[test]
    fn test_protocol_duality_symmetry() {
        // This test verifies the symmetry of the duality relationship
        // by checking that the dual of the dual is the original type.
        
        // We can't directly compare types in Rust, but we can use type inference
        // to verify that the types match.
        
        // TestProtocol::Dual is DualTestProtocol
        // DualTestProtocol::Dual is TestProtocol
        // So TestProtocol::Dual::Dual is TestProtocol, which confirms the symmetry property.
        
        // This is a compile-time check - if these types match up correctly,
        // the code will compile successfully.
        fn check_dual_symmetry<P, Q>()
        where
            P: Protocol<Dual = Q>,
            Q: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // Verify that TestProtocol and DualTestProtocol are duals of each other
        check_dual_symmetry::<TestProtocol, DualTestProtocol>();
    }
}