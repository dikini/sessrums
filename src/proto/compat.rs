//! Compatibility layer between binary and multiparty session types.
//!
//! This module provides types and traits to ensure backward compatibility
//! between binary session types and multiparty session types (MPST).

use crate::proto::Protocol;
use crate::proto::roles::Role;
use std::marker::PhantomData;

/// A trait for converting between binary and multiparty session types.
///
/// This trait allows for seamless integration between the existing binary
/// session types and the new multiparty session types.
pub trait ProtocolCompat<R: Role> {
    /// The binary protocol type that is compatible with this multiparty protocol.
    type BinaryProtocol: Protocol;
    
    /// Converts a multiparty protocol to a binary protocol.
    fn to_binary(self) -> Self::BinaryProtocol;
    
    /// Converts a binary protocol to a multiparty protocol.
    fn from_binary(binary: Self::BinaryProtocol) -> Self;
}

/// A wrapper for binary session types to make them compatible with MPST.
///
/// This wrapper allows existing binary session types to be used in MPST contexts.
pub struct BinaryWrapper<P: Protocol, R: Role>(P, PhantomData<R>);

impl<P: Protocol, R: Role> BinaryWrapper<P, R> {
    /// Creates a new binary wrapper for the given protocol and role.
    pub fn new(protocol: P) -> Self {
        BinaryWrapper(protocol, PhantomData)
    }
    
    /// Unwraps the binary protocol.
    pub fn unwrap(self) -> P {
        self.0
    }
}

impl<P: Protocol, R: Role> Protocol for BinaryWrapper<P, R> {
    type Dual = BinaryWrapper<P::Dual, R>;
}

/// A wrapper for MPST local types to make them compatible with binary session types.
///
/// This wrapper allows MPST local types to be used in binary session type contexts.
pub struct MPSTWrapper<P, R: Role>(P, PhantomData<R>);

impl<P, R: Role> MPSTWrapper<P, R> {
    /// Creates a new MPST wrapper for the given protocol and role.
    pub fn new(protocol: P) -> Self {
        MPSTWrapper(protocol, PhantomData)
    }
    
    /// Unwraps the MPST local protocol.
    pub fn unwrap(self) -> P {
        self.0
    }
}

impl<P, R: Role> Protocol for MPSTWrapper<P, R>
where
    P: 'static
{
    type Dual = MPSTWrapper<P, R>;
}

// Implement ProtocolCompat for common protocol types
// These implementations allow for seamless conversion between binary and multiparty protocols

// Implementation for Send
impl<T, P: Protocol, R: Role> ProtocolCompat<R> for crate::proto::send::Send<T, P> {
    type BinaryProtocol = Self;
    
    fn to_binary(self) -> Self::BinaryProtocol {
        self
    }
    
    fn from_binary(binary: Self::BinaryProtocol) -> Self {
        binary
    }
}

// Implementation for Recv
impl<T, P: Protocol, R: Role> ProtocolCompat<R> for crate::proto::recv::Recv<T, P> {
    type BinaryProtocol = Self;
    
    fn to_binary(self) -> Self::BinaryProtocol {
        self
    }
    
    fn from_binary(binary: Self::BinaryProtocol) -> Self {
        binary
    }
}

// Implementation for End
impl<R: Role> ProtocolCompat<R> for crate::proto::end::End {
    type BinaryProtocol = Self;
    
    fn to_binary(self) -> Self::BinaryProtocol {
        self
    }
    
    fn from_binary(binary: Self::BinaryProtocol) -> Self {
        binary
    }
}

// Implementation for Choose
impl<L: Protocol, R1: Protocol, R2: Role> ProtocolCompat<R2> for crate::proto::choose::Choose<L, R1> {
    type BinaryProtocol = Self;
    
    fn to_binary(self) -> Self::BinaryProtocol {
        self
    }
    
    fn from_binary(binary: Self::BinaryProtocol) -> Self {
        binary
    }
}

// Implementation for Offer
impl<L: Protocol, R1: Protocol, R2: Role> ProtocolCompat<R2> for crate::proto::offer::Offer<L, R1> {
    type BinaryProtocol = Self;
    
    fn to_binary(self) -> Self::BinaryProtocol {
        self
    }
    
    fn from_binary(binary: Self::BinaryProtocol) -> Self {
        binary
    }
}

// Implementation for Rec
impl<P: Protocol, R: Role> ProtocolCompat<R> for crate::proto::rec::Rec<P> {
    type BinaryProtocol = Self;
    
    fn to_binary(self) -> Self::BinaryProtocol {
        self
    }
    
    fn from_binary(binary: Self::BinaryProtocol) -> Self {
        binary
    }
}

// Implementation for Var
impl<const N: usize, R: Role> ProtocolCompat<R> for crate::proto::var::Var<N> {
    type BinaryProtocol = Self;
    
    fn to_binary(self) -> Self::BinaryProtocol {
        self
    }
    
    fn from_binary(binary: Self::BinaryProtocol) -> Self {
        binary
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::proto::send::Send;
    use crate::proto::recv::Recv;
    use crate::proto::end::End;
    use crate::proto::roles::{RoleA, RoleB};
    
    #[test]
    fn test_binary_wrapper() {
        // Create a binary protocol
        type BinaryProtocol = Send<i32, Recv<String, End>>;
        
        // Wrap it for use with MPST
        let wrapped = BinaryWrapper::<BinaryProtocol, RoleA>::new(Default::default());
        
        // NOTE: protocol_name assertion removed as it's not part of the wrapper API
    }
    
    #[test]
    fn test_mpst_wrapper() {
        // Create an MPST local protocol (using binary types for simplicity in this test)
        type MPSTLocalProtocol = Send<i32, Recv<String, End>>;
        
        // Wrap it for use with binary session types
        let wrapped = MPSTWrapper::<MPSTLocalProtocol, RoleA>::new(Default::default());
        
        // NOTE: protocol_name assertion removed as it's not part of the wrapper API
    }
    
    #[test]
    fn test_protocol_compat() {
        // Test conversion for Send
        let send_protocol: Send<i32, End> = Default::default();
        let binary = <Send<i32, End> as ProtocolCompat<RoleA>>::to_binary(send_protocol); // Specify RoleA
        let _mpst = <Send<i32, End> as ProtocolCompat<RoleA>>::from_binary(binary); // Specify RoleA
        
        // Test conversion for Recv
        let recv_protocol: Recv<String, End> = Default::default();
        let binary = <Recv<String, End> as ProtocolCompat<RoleA>>::to_binary(recv_protocol); // Specify RoleA
        let _mpst = <Recv<String, End> as ProtocolCompat<RoleA>>::from_binary(binary); // Specify RoleA
        
        // Test conversion for End
        let end_protocol: End = Default::default(); // End is an empty struct, Default works
        let binary = <End as ProtocolCompat<RoleA>>::to_binary(end_protocol); // Specify RoleA
        let _mpst = <End as ProtocolCompat<RoleA>>::from_binary(binary); // Specify RoleA, assign to _mpst
    }
}