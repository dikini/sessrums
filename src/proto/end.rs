//! End protocol type for session types.
//!
//! This module defines the `End` protocol type, which represents the end of a
//! communication session. It is a terminal protocol that indicates no further
//! communication will occur.

use super::Protocol;

/// A protocol that represents the end of a communication session.
///
/// The `End` protocol is a terminal protocol that indicates no further
/// communication will occur. It is typically used as the final protocol
/// in a chain of protocol compositions.
///
/// # Duality
///
/// The dual of `End` is `End` itself, as ending a session is symmetric
/// for both parties involved in the communication.
///
/// # Examples
///
/// ```
/// use sessrums::proto::{Protocol, Send, End};
///
/// // A protocol that sends an i32 and then ends
/// type SimpleProtocol = Send<i32, End>;
/// ```
#[derive(Debug)]
pub struct End;

impl Protocol for End {
    /// The dual of `End` is `End` itself.
    ///
    /// This reflects the fact that ending a session is symmetric
    /// for both parties involved in the communication.
    type Dual = End;
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::proto::{Send, Recv};

    #[test]
    fn test_end_protocol_implementation() {
        // This test verifies that End implements Protocol
        // and that its Dual is defined as End itself
        
        // We can't directly test the equality of types in Rust,
        // but we can verify that End implements Protocol
        fn assert_implements_protocol<T: Protocol>() {}
        
        // This will compile only if End implements Protocol
        assert_implements_protocol::<End>();
        
        // We can also verify that the associated type Dual is defined correctly
        fn assert_dual_type<T, D>()
        where
            T: Protocol<Dual = D>,
        {}
        
        // This will compile only if End::Dual is End
        assert_dual_type::<End, End>();
    }

    #[test]
    fn test_end_duality_symmetry() {
        // This test verifies the symmetry of the duality relationship for End
        // End::Dual is End, and End::Dual::Dual is End
        
        // Check that End::Dual is End and End::Dual::Dual is End
        fn check_duality<P, D>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if the duality relationship is correctly implemented
        check_duality::<End, End>();
    }

    #[test]
    fn test_end_with_send_composition() {
        // Test End in composition with Send
        
        // A protocol that sends an i32 and then ends
        type SendThenEnd = Send<i32, End>;
        
        // The dual should be Recv<i32, End>
        type ExpectedDual = Recv<i32, End>;
        
        // Check that the dual relationships are correctly established
        fn check_dual<P, D>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if SendThenEnd's dual is correctly derived
        check_dual::<SendThenEnd, ExpectedDual>();
    }

    #[test]
    fn test_end_with_recv_composition() {
        // Test End in composition with Recv
        
        // A protocol that receives an i32 and then ends
        type RecvThenEnd = Recv<i32, End>;
        
        // The dual should be Send<i32, End>
        type ExpectedDual = Send<i32, End>;
        
        // Check that the dual relationships are correctly established
        fn check_dual<P, D>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if RecvThenEnd's dual is correctly derived
        check_dual::<RecvThenEnd, ExpectedDual>();
    }

    #[test]
    fn test_end_with_complex_composition() {
        // Test End in a more complex protocol composition
        
        // A protocol that sends an i32, receives a String, and then ends
        type ComplexProtocol = Send<i32, Recv<String, End>>;
        
        // The dual should be Recv<i32, Send<String, End>>
        type ExpectedDual = Recv<i32, Send<String, End>>;
        
        // Check that the dual relationships are correctly established
        fn check_complex_dual<P, D>()
        where
            P: Protocol<Dual = D>,
            D: Protocol<Dual = P>,
        {
            // Empty function body - we're just checking type relationships
        }
        
        // This will compile only if ComplexProtocol's dual is correctly derived
        check_complex_dual::<ComplexProtocol, ExpectedDual>();
    }
}