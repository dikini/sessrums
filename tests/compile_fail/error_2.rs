//! Error Example 2: Deadlock (Send/Send)
//!
//! This example demonstrates a protocol that should fail to compile because
//! both client and server try to send an i32 first, which would cause a deadlock.
//!
//! # Why This Protocol Fails
//!
//! In session types, communication between two parties must be complementary to avoid deadlocks.
//! When one party sends a message, the other party must receive a message. This complementary
//! relationship is formalized through the concept of "duality".
//!
//! In this example:
//! - Client protocol: Send<i32, End> (send an i32, then end)
//! - Server protocol: Send<i32, End> (send an i32, then end)
//!
//! Both parties are trying to send a message, but neither is receiving one. This creates a
//! deadlock situation where both parties might block trying to send messages that will
//! never be received.
//!
//! # How Session Types Prevent This Error
//!
//! The session type system prevents this error at compile time by enforcing duality between
//! communicating parties. For any protocol P, there must exist a dual protocol P::Dual that
//! represents the complementary behavior.
//!
//! For Send<T, P>, the dual is Recv<T, P::Dual>. This means:
//! - The dual of Send<i32, End> is Recv<i32, End>
//! - The dual of Recv<i32, End> is Send<i32, End>
//! - The dual of End is End (termination is symmetric)
//!
//! When we try to create a channel pair with non-dual protocols, the type system rejects it,
//! preventing the deadlock at compile time.
//!
//! # Visual Representation of the Protocol
//!
//! ```text
//! Client                 Server
//!   |                      |
//!   |  !i32                |  !i32
//!   | <---- DEADLOCK ----> |
//!   |                      |
//!  End                    End
//! ```
//!
//! Legend:
//! - ?T: Receive a value of type T
//! - !T: Send a value of type T
//! - DEADLOCK: Both parties trying to send, neither receiving
//!
//! # Correct Version (For Reference)
//!
//! A correct version would have complementary actions:
//!
//! ```text
//! Client                 Server
//!   |                      |
//!   |  !i32                |  ?i32
//!   | -------------------> |
//!   |                      |
//!  End                    End
//! ```

use sez::proto::{Protocol, Send, End};
use sez::chan::Chan;

// Define the protocol types
type ClientProto = Send<i32, End>;
type ServerProto = Send<i32, End>; // Not the dual of ClientProto! Should be Recv<i32, End>

// This function attempts to create a pair of channels with non-dual protocols
fn create_invalid_channel_pair() {
    // This should fail to compile because ServerProto is not the dual of ClientProto
    let client_chan = Chan::<ClientProto, _>::new(());
    let server_chan = Chan::<ServerProto, _>::new(());
    
    // In a real implementation, we would use a function like:
    // let (client_chan, server_chan) = session_channel::<ClientProto, ServerProto>((), ());
    // But that would enforce duality, so we're manually creating the channels here
    // to demonstrate the error.
    
    // The type system should prevent this code from compiling because
    // the protocols are not duals of each other.
    
    // This function would verify that the protocols are duals of each other
    verify_dual_protocols(client_chan, server_chan);
}

// This function verifies that two protocols are duals of each other
// It should fail to compile if P::Dual != Q
fn verify_dual_protocols<P, Q, IO1, IO2>(_chan1: Chan<P, IO1>, _chan2: Chan<Q, IO2>)
where
    P: Protocol,
    Q: Protocol,
    P::Dual: Protocol<Dual = P>,
    Q: Protocol<Dual = P>  // This is the key constraint that should fail
{
    // The function body doesn't matter, as the type constraints will cause a compile error
    // if the protocols are not duals of each other
    
    // In a real session type system, this verification would happen automatically
    // when creating a channel pair, ensuring that the two endpoints have compatible
    // protocols that won't deadlock.
}

fn main() {
    create_invalid_channel_pair();
    
    // If this code were to compile and run (which it shouldn't),
    // the following would happen at runtime:
    
    // 1. Client would try to send an i32
    // 2. Server would also try to send an i32
    // 3. Depending on the underlying channel implementation:
    //    - Both might block forever if the channels are synchronous
    //    - The channels might fill up if they're buffered
    //    - The messages might be lost if no one is receiving
    
    // Fortunately, the session type system prevents this at compile time
    // by enforcing duality between the client and server protocols.
}