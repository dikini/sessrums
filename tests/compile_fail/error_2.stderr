error[E0271]: type mismatch resolving `<Send<i32, End> as Protocol>::Dual == Send<i32, End>`
  --> tests/compile_fail/error_2.rs:85:40
   |
85 |     verify_dual_protocols(client_chan, server_chan);
   |     ---------------------              ^^^^^^^^^^^ expected `Send<i32, End>`, found `Recv<i32, End>`
   |     |
   |     required by a bound introduced by this call
   |
   = note: expected struct `sessrums::proto::Send<i32, sessrums::proto::End>`
              found struct `Recv<i32, sessrums::proto::End>`
note: required by a bound in `verify_dual_protocols`
  --> tests/compile_fail/error_2.rs:95:17
   |
90 | fn verify_dual_protocols<P, Q, IO1, IO2>(_chan1: Chan<P, IO1>, _chan2: Chan<Q, IO2>)
   |    --------------------- required by a bound in this function
...
95 |     Q: Protocol<Dual = P>  // This is the key constraint that should fail
   |                 ^^^^^^^^ required by this bound in `verify_dual_protocols`
